[레벨 1 - 자동차경주] 조이썬(이영수) 미션 제출합니다.

youngsu5582 opened the pull request:
# 자동차 경주

사용자가 콘솔을 통해
자동차 이름 과 경기 진행 횟수를 입력해
경기 진행 상태 와 경기 결과를 출력하는 게임입니다

페어 ( 폰드 - tackyu ) 와 함께 진행했습니다

## 진행중 느낀 점

페어 프로그래밍 이라는 것을 처음 해본 기회였는데,
정말 색다른 경험 이였습니다.

서로 다른 세팅 을 가진 맥북에서 협업점을 찾기 위해서
Code With Me 를 알고 연결한 후
서로 간 다른 코드 스타일 및 개성 까지
혼자서 코드를 짠 저에게 페어 프로그래밍이란 참 어려웠습니다

처음에는 시간을 잘 못 지키기도 하고
서로 간 역활 을 교체하며 Context Switching 도 문제 였습니다

서로 개발 효율을 증진시키기 위해 노력했고 다양한 방법을 시도해봤습니다
저희가 느낀 최고의 효율은 끊임없이 얘기하고 , 끊임없이 토론한다 였던거 같습니다

서로가 작업간 공백을 끊기지 않게 하기 위해
서로가 다른 코드 스타일을 맞추기 위해
조율 해나가는 과정은 정말 재밌었습니다

둘째날인 오늘 목요일은 서로가 간단한 얘기를 하면서 코드를 짜도
Context Switching 이 발생하지 않으며 무사히 완수 했던 거 같습니다

그럼에도 , 서로 간 다른 코드 스타일 과정 중
발생한 코드 스타일 불일치나 시간 부족으로 인한 상수 및 리팩토링 부족등에서 아쉬움을 느꼈습니다

이번 스텝을 통해 다음에는 좀더 우아한 뛰어난 페어 프로그래밍을 해보겠다고 다짐했습니다
Thanks to 우테코 , Thanks to 폰드!

## 진행 중 궁금했던 점

이번 과제 구현 중 도메인을 많이 생성하고
도메인을 통해 기능을 구현하려고 집중을 했습니다.

이 와중에 Controller 나 Service 같은 클래스가 처리해도 되는 로직을
도메인에서 처리하게 했나 걱정도 조금 됩니다..!
( 해당 부분에 대해서도 피드백 주시면 정말 감사하겠습니다! )

아래는 궁금한 내용들입니다

### 일급 객체

저희는 단순 값을 가지는 일급 객체를 많이 만들어서 사용했습니다
```java
public class Cars {
    private final List<Car> value;
    ...
}

public class Name {
      private String value;
      public Name(String value) {
          validate(value);
          this.value = value;
  }
}
```
validate 를 하기 위해서 의 일급 객체는
도메인 내 추가 검증을 위해 필요하다고 생각하나,
단순히 List<Car> 를 감싸기 위한 일급 객체가 필요한지에 대해 궁금했습니다.

그리고 이러한 일급 객체가 필요하다고 생각하면
메소드의 파라미터에
1. getValue() 를 통해 List<Car> ...
2. Cars 를 통해 넘긴 후 메소드 내부에 getValue()

두 가지중 어떤게 맞는 방법인지에 대해도 궁금합니다! 🫡 🫡

### 단순 객체

```java
public class RaceProgress {
    private final List<RaceCarInfo> raceCarInfos;
    private RaceProgress(List<RaceCarInfo> raceCarInfos){
        this.raceCarInfos = raceCarInfos;
    }
    public static RaceProgress from (final List<Car> cars){
        return new RaceProgress(cars
                .stream()
                .map(RaceCarInfo::from)
                .toList()
        );
    }
    @Override
    public String toString() {
        return raceCarInfos
                .stream()
                .map(RaceCarInfo::toString)
                .collect(Collectors.joining("\n","","\n"));
    }
}
```

해당 RaceProgress 는 
각 순차 마다 자동차들의 진행 상태를 출력해주기 위해
Car -> RaceCarInfo 로 변환 후
해당 RaceCarInfo 를 List 로 가지는 객체입니다

해당 객체는 단순히 객체 지향적 설계 와 출력을 편하게 하기 위해 만들어진 객체입니다

제 생각으로는 이런 객체가
일종의 중간자 역활 의 객체라고 생각하는데,

해당 객체가 굳이 필요한지에 대해서 궁금합니다!

또한 , List<Car> 라는 다른 객체에서 -> RaceCarInfo 를 가지는 RaceProgress
객체 설계 순서가 맞는 설계 방식인지도 궁금합니다!

이상입니다 감사합니다 🙏🙏

tackyu committed:
docs: 기능 명세서 초안 정리

tackyu committed:
feat : Name Domain 생성

이에 따른 테스트 코드 추가

tackyu committed:
feat : RaceCount Domain 생성

이에 따른 테스트 코드 추가

tackyu committed:
feat : Car Domain 생성

이에 따른 테스트 코드 추가

tackyu committed:
feat : RaceResult Domain 생성

이에 따른 테스트 코드 추가

tackyu committed:
feat : Car Domain 생성

이에 따른 테스트 코드 추가

tackyu committed:
feat : NumberGenerator Util 생성

이에 따른 테스트 코드 추가

tackyu committed:
feat : Splitter Util 생성

이에 따른 테스트 코드 추가

tackyu committed:
feat : InputView View 생성

tackyu committed:
feat : OutputView View 생성

tackyu committed:
feat : Cars Domain 생성

이에 따른 테스트 코드 추가

tackyu committed:
feat : RaceCarInfo Domain 생성

이에 따른 테스트 코드 추가

tackyu committed:
feat : RaceProgress Domain 생성

이에 따른 테스트코드 작성

tackyu committed:
feat : Name Domain getValue 메서드 생성

tackyu committed:
feat : Car Domain에 전진 숫자 최대값 추가

tackyu committed:
feat : 게임 로직을 담당하는 GameController 생성

tackyu committed:
feat : GameController 통해 게임을 진행하는 Main 메서드 구현

tackyu committed:
refactor : Domain 패키지 분리

tackyu committed:
docs : 기능 구현 완료에 따른 기능 명세서 추가 수정

lxxjn0 commented on docs/README.md:
```diff
@@ -0,0 +1,43 @@
+# java-racingcar
+
+## 개발 기능 목록
```
개발 기능 목록을 상세하게 작성해주셨네요. 커밋도 적절한 단위로 잘 작성해주신 것 같아요 :) 그런데 조이썬이 작성해주신 문서가 기능 목록이기도 하지만, 그보다는 코드 명세서? 설명서에 더 가까운 것 같아요. 

미션에서 요구하는 기능 구현 목록 작성은 이번 미션의 요구사항을 통해서 어떠한 기능이 구현되어야 하는지를 나열해보고 개발의 방향을 잡기 위함이라고 생각해주시면 좋을 것 같습니다. 스스로의 개발 편의를 위한 장치이기에 앞으로 미션을 하시면서 더 적절한 방향을 찾아보는 것도 좋을 것 같습니다 👍🏼

lxxjn0 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,15 @@
+public class Main {
+
+    public static void main(String[] args) {
+        InputView inputView=new InputView();
+        OutputView outputView=new OutputView();
```
코드 컨벤션이 올바르게 적용되지 못한 것 같아요. 구글 자바 코딩 컨벤션을 확인하고 적용해보시면 좋을 것 같습니다.

- https://myeonguni.tistory.com/1596
- https://tychejin.tistory.com/334

lxxjn0 commented on src/main/java/domain/name/NameCatalog.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class NameCatalog {
```
[제안] NameCatalog라는 이름이 잘못되었다는 것은 아니지만 보편적으로 일급 컬렉션은 컬렉션이 가지는 객체의 복수형을 쓰거나 또는 비즈니스 로직에서 의미가 있는 이름을 주로 활용하는 것 같습니다. 지금과 같은 경우에는 Names 등이 예시가 될 수 있을 것 같아요.

추가로 catalog라는 단어가 단순히 목록을 뜻하기 보다는 일람표, 상품안내서, 홍보물 등으로 주로 사용되는 단어인 것 같아요. 따라서 조금은 적절한 이름이 아닐 수도 있다는 생각이 들어 코멘트 남겨봤습니다.

youngsu5582 replied:
넵!
Cars 는 s 잘 붙혀놓고
Name 에는 "이름을 저장한 목록을 나타내야겠다" 에 꽂혀서
오히려 더 알아보기 힘들고 복잡한 네이밍 이 탄생한거 같아요!

lxxjn0 commented on src/main/java/domain/name/NameCatalog.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    private void validate(List<Name> names){
+        Integer size=names.size();
```
[제안] Java에서 자료형은 크게 기본형과 참조형이 있습니다. 그리고 기본형을 표현할 수 있는 primitive type이 있고, 이 기본형을 참조할 수 있게 객체로 감싼 wrapper class가 존재합니다. 여기서 조이썬이 사용한 Integer는 기본형을 감싼 wrapper class 입니다.

primitive type과 wrapper class의 차이가 무엇인지, 각각을 어떠한 상황에서 사용하는 것이 적절할 지 한번 공부해보시면 좋을 것 같습니다.

youngsu5582 replied:
저는 항상
숫자 뿐만 아니라 , 
참조형 자료형이 가독성이 좋고 통일성 때문에 좋다고 생각을 해서 기본형을 사용하지 않고 참조형만 사용했습니다
해당 제안을 통해 찾아보았습니다. ( 연극 준비 때문에 시간 부족으로 간단하게만 한 거 같습니다,,, )

### 기본형 (Primitive Types)

- 메모리 효율성이 높고 , 처리 속도 빠름
- 스택 메모리에 저장 , 실제 값 저장
- null 값 불가

### 참조형

- 힙 메모리 저장 , 자료가 저장된 주소 저장
- 주소를 참조해 
- null 가능

즉 , 간단한 연산 & 성능을 위해서는 기본형을 사용 +
AutoBoxing 가 있기는 하나 , 불필요한 연산 발생 ( List 의 size 는 int 로 반환 )

이러한 상황을 통해 int 가 더 적절하다는 것을 깨달은거 같습니다.

여기서 , 추가적인 궁금함 으로
도메인 설계할 때도
위의 기본형 특징 3개를 만족한다고 하면
int 로 멤버 변수를 지정해도 괜찮나요??

lxxjn0 replied:
넵!! 자료형의 선택은 개발자의 판단에 따라 선택하시면 됩니다. 필드로 기본 자료형을 가지고 있는 것은 전혀 어색하지 않습니다. 이 기본 자료형을 코드에서 공통적으로 활용하기 위해서 어쩔 수 없이 참조 자료형으로 사용되곤 하는데(예. Stream 메서드) 이처럼 상황과 사용 용도에 따라서 결정하고 사용하시면 좋을 것 같습니다.

lxxjn0 commented on src/main/java/domain/name/NameCatalog.java:
```diff
@@ -0,0 +1,33 @@
+    private void validate(List<Name> names){
+        Integer size=names.size();
+
+        Set<String> nonDuplicate = names
+                .stream()
+                .map(Name::getValue)
+                .collect(Collectors.toSet());
```
Java의 equals와 hashCode를 사용하면 Name 객체에서 이름을 꺼내지 않고도 해결할 수 있을 것 같습니다.

youngsu5582 replied:
저도 해당 부분이 불필요한 , 작업을 하는거 같아서
계속 신경이 쓰였는데 , 
equals 를 오버라이딩 해 검증 시키면 되겠네요 🙏🙏

lxxjn0 commented on src/main/java/domain/car/Car.java:
```diff
@@ -0,0 +1,33 @@
+import domain.name.Name;
+
+public class Car  {
+    public static final Integer FORWARD_MAX_NUMBER = 10;
```
[질문] 해당 상수를 Car 클래스에서 관리하신 이유가 무엇인가요?

youngsu5582 replied:
해당 상수는
차가 전진을 결정하는 숫자들 중 가장 큰 값을 의미하기 때문에
Car 가 가지고 있다고 판단하여 클래스 내에 선언을 했었습니다.
이때도 , 상수를 만들면서
public 인 상수는 처음 만들어서 해당 내용이 맞나 생각을 했었습니다.

질문을 받으며 다시 생각하며 느낀점은
해당 숫자가 차가 가질 수 있는 속도의 최대값 이더라도
해당 도메인에서 사용하지 않고
도메인이 상수를 제공해주는 방식이
객체지향적 설계이지 않은거 같다고 느꼈습니다

RaceCarConstant 와 같은 클래스를 만들어
상수를 관리하게 할 거 같습니다.

감사합니다 ☺️

lxxjn0 replied:
도메인에서 관리해야 하는 상수를 외부에 제공하는 것은 적절하다고 생각합니다. 제가 생각하는 자동차의 책임과 역할은 '자동차의 이름과 위치를 관리하며 게임 진행에 따라 전진 또는 정지를 한다'라고 생각됩니다. 

요구사항 중 '랜덤한 값은 0부터 9 사이의 숫자로 생성한다'는 내용은 자동차 도메인보다는 자동차 게임을 관리하는 도메인에 적절하다는 생각이 들어서 이 부분에 대해 질문을 드렸던 것이었습니다.

이번에 상수만 관리하기 위한 클래스를 별도로 만들었는데 제가 느끼긴 오히려 이 부분이 좀 더 객체지향적이지 않다고 생각합니다. 관리해야 하는 도메인을 담고있는 클래스가 있고, 여기서 관리해야 할 정보 중 상수도 포함되기 때문에 적절한 도메인 객체/클래스에 상수가 함께 관리되어야 이후 변경 범위가 도메인 단위로 제한될 수 있을 것 같아요.

이번 자동차 게임 요구사항에서 어떠한 부분을 각각의 도메인으로 규정할 것인지, 어떤 클래스에서 어떻게 도메인을 코드로 관리할 것인지 등을 한번 더 고민해보면서 2단계 진행해보시면 많은 도움이 될 수 있을 것 같습니다.

코드를 코드로만 바라보기 보다는 비즈니스 로직을 코드로 옮긴다는 관점에서 바라보신다면 오히려 클래스 구분과 관련 코드 작성이 쉬우실 수도 있을 것 같습니다 :)

youngsu5582 replied:
단순히 크게 자동차 도메인이 가지고 있는게 적절하다 했는데
자동차 게임 <-> 자동차
이렇게 도메인을 분리하고 좀더 상수를 세밀화 할 수 있는거 같습니다!

해당 부분은 프리코스 할때도 계속 생각했던 내용인데 , 설명해주신 내용이 맞는거 같습니다
상수를 Constant 파일로 분리하며
결국 도메인이 상수 관리해야 하는 파일에 의존 해야하고
불필요한 클래스가 늘어난다고 생각했었습니다

좀더 넓고 비즈니스 로직 적인 관점에서 바라보도록 해보겠습니다! 감사합니다!!

lxxjn0 commented on src/main/java/domain/car/Car.java:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    @Override
+    public String toString() {
+        return name.getValue();
+    }
```
Java에서 toString의 용례에 대해서 한번 알아보시면 좋을 것 같습니다. 그리고 현재 toString을 사용하는 상황이 적절한지 한번 고민해보시면 좋을 것 같아요.

youngsu5582 replied:
[Effective Java) Item 12. toString을 항상 재정의 하라
 ](https://medium.com/@bnm1128/effective-java-item-12-tostring%EC%9D%84-%ED%95%AD%EC%83%81-%EC%9E%AC%EC%A0%95%EC%9D%98-%ED%95%98%EB%9D%BC-8723f9d4f35f)

해당 블로그에 잘 정리되어 있어 찾아 봤습니다!

> toString 은 간결하며 사람이 읽기 쉬운 형태 유익한 정보 반환해야 한다

#### 좋은 toString 재정의 방법

- 객체가 가진 주요 정보 모두 반환 ( 거대하다면 요약 정보 )
- 반환값 포맷을 문서화 여부 결정

여기서 사용한 toString 은
두가지 방법을 다 만족하지 못하는
OutputView 에서 출력을 용이하기 위해 만든 Bad Code 임을 깨달았습니다 🥹
고민해보라고 해주셔서 감사합니다!

lxxjn0 commented on src/main/java/domain/race/RaceCarInfo.java:
```diff
@@ -0,0 +1,24 @@
+        return new RaceCarInfo(car.getName(),car.getPosition());
+    }
+
+    @Override
+    public String toString() {
+        return name.getValue() + " : " + "-".repeat(position);
+    }
```
앞선 피드백에서 toString에 대해 코멘트 남겼는데 추가적인 설명이 있으면 좋을 것 같아 코멘트 남깁니다. 현재 toString 메서드는 자동차의 위치를 `-`으로 나타내는 메서드로 보여집니다. 

그런데 만약 요구사항에서 '자동차의 위치를 *로 출력한다'라고 변경이 된다면 어떻게 될까요? 해당 요구사항은 출력하는 방법에 대한 변화이므로 입출력 로직 변경으로 보여집니다. 하지만 현재 자동차의 위치는 도메인 객체에서 관리하고 있으므로 코드 상에서는 입출력 로직 변경이 아닌 도메인 로직 변경에 해당되게 됩니다. 이런 관점을 예시로 생각해보면서 코드를 한번 리팩토링 해보시면 좋을 것 같습니다 :)

youngsu5582 replied:
해당 부분도 
toString 의 용례 ,
그리고 OutputView 에서 해야하는 로직을
편의성을 위해 도메인에서 수행한다는 것을 깨달았습니다.

해당 관점을 시작으로 리팩토링 해보겠습니다!

lxxjn0 commented on src/main/java/domain/race/RaceResult.java:
```diff
@@ -0,0 +1,35 @@
+        List<Car> winnerCars = selectWinnerCars(cars,maxPosition);
+        return new RaceResult(winnerCars);
+    }
+    private static Integer getMaxPosition(List<Car> cars){
+        return cars.
+                stream()
+                .mapToInt(Car::getPosition)
+                .max()
+                .orElse(0);
+    }
+    private static List<Car> selectWinnerCars(List<Car>cars,Integer maxPosition){
+        return cars
+                .stream()
+                .filter(car -> car.getPosition().equals(maxPosition))
+                .toList();
+    }
```
이러한 로직을 보면 cars 리스트를 사용해서 로직을 전개하고 있습니다. 앞서 Cars라는 일급 컬렉션을 생성하였는데 이 일급 컬렉션을 활용해보는 것은 어떨까요?

youngsu5582 replied:
Cars 가 단순히 car 의 배열만 가진게 아닌,
car 배열들을 활용하는 기능들을 제공해주는게 맞겠네요!

```java
    public List<Car> getMaxPositionCars(){
        Integer maxPosition = getMaxPosition();
        return selectCarsWithPosition(maxPosition);
    }
    private Integer getMaxPosition(){
        return value.
                stream()
                .mapToInt(Car::getPosition)
                .max()
                .orElse(0);
    }
    private List<Car> selectCarsWithPosition(Integer position){
        return value
                .stream()
                .filter(car -> car.getPosition().equals(position))
                .toList();
    }
```

근데 궁금한게 
이렇게 멤버 변수로 LIst<Car> 만 가지는게 자명한 경우에는
```java
    private Stream<Car> stream(){
        return value.stream();
    }
    private List<Car> selectCarsWithPosition(Integer position){
        return
                 stream()
                .filter(car -> car.getPosition().equals(position))
                .toList();
    }
```

이런식으로
value.stream() -> stream() 으로 변경하는건 굳이 불필요한 코드인지 궁금합니다!

lxxjn0 replied:
단순히 배열을 스트림으로 바꾸기 위한 코드라면 꼭 분리를 할 필요는 없을 것 같습니다. 물론 private 메서드이기 때문에 메서드 분리에 대한 관점을 조금은 여유있게 가져가도 되긴 합니다.(외부에 노출이 되지 않고 해당 클래스에서만 사용되기 때문에)

저 같은 경우에는 반복을 제거하기 위한 용도로 메서드를 사용하긴 하지만 그보다는 메서드로서의 의미가 있는지를 먼저 고려해보려고 합니다. 단순히 스트림으로 변환하는 역할은 메서드의 존재 이유가 되기엔 사소하다고 생각이 되어서 저 같은 경우에는 분리하지 않을 것 같아요 :)

lxxjn0 commented on src/main/java/util/NumberGenerator.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.Random;
+
+public class NumberGenerator {
+    public static Integer generate(final Integer value){
+        return new Random().nextInt(value);
+    }
+}
```
[질문] 해당 클래스는 사실 generate 메서드를 호출하는 로직을 new Random().nextInt() 메서드를 호출하는 로직으로 바꿔도 전혀 지장이 없는 메서드로 보여집니다. 해당 클래스를 어떠한 기준을 통해서 만드셨나요?

youngsu5582 replied:
해당 클래스는
race 메소드 내에서 , 
관련 없는 new Random().nextInt() 가 아닌 
1. 명확하게 숫자를 생성한다는 의미를 주기 위해
2. Controller 가 new Random().nextInt() 를 가지는 것이 이상하다는 판단

라는 기준을 가지고 진행을 했었습니다!
하지만 , 다시 보니 NumberGenerator 를 사용하나 , 안 사용하나 큰 차이가 없는거 같습니다,,

추가적인 질문으로는 ,
```java
public interface NumberGenerator {
    abstract Integer generate(final Integer value);
}
```

```java
public class RandomNumberGenerator implements NumberGenerator{
    public Integer generate(final Integer value){
        return new Random().nextInt(value);
    }
}
```
기존의 코드를 
테스트에서 사용하기 위해서 ( MockingNumberGenerator : 숫자를 순서대로 ... ) 와 
NumberGenerator 를 교체하기 용이하기 쉬울 거 같다는 생각은 했는데,

이 방법 역시도 , 
스티치가 말한대로 new Random().nextInt() 의 코드를 바꿀만큼의
가치가 있다고는 생각이 들지 않는거 같아요..!

제 솔직한 마음으로는 , 그래도 new Random().nextIn() 가 있는것 보단 
NumberGenerator Class 를 가지고 있는데 맞다고 생각하는데
힌트나 의견이 가능할까요...? 😭😭

lxxjn0 replied:
조이썬의 의견을 잘 말씀해주셔서 감사합니다 :)

사소한 부분일 수 있지만 만약 해당 코드가 아래와 같다면 해당 메서드의 존재에 대한 질문을 하지 않았을 것 같아요.

```java
public class RandomNumberGenerator {

    private static final int RANDOM_NUMBER_UPPER_BOUND = 10;

    public int generate() {
        return new Random().nextInt(RANDOM_NUMBER_UPPER_BOUND);
    }
}
```

이 코드를 보면 자동차 게임에서 '0에서 9 사이의 랜덤한 숫자를 생성한다'라는 비즈니스 로직을 구현한 클래스로 볼 수 있습니다. 온전히 해당 비즈니스만을 제공하는 코드가 될 수 있기 때문입니다.

이전에 조이썬이 작성한 코드는 단순한 유틸리티성 코드이고, 그 로직 또한 굳이 클래스로 관리할 만큼의 로직이 아니기에 질문을 드린 것이었어요.

조이썬의 질문대로 지금의 코드가 존재함으로 테스트를 하기에 용이한 코드가 될 수 있습니다. 단지 메서드의 존재 이유를 묻는다면 합당한 이유가 보이지 않았기에 질문을 드린 것이었습니다.

제 답변이 조이썬의 질문에 대한 답이 되었을지 모르겠지만 혹시 이해가 안가는 부분이 있다면 언제든 다시 질문주세요 :)

youngsu5582 replied:
해당 기능 자체 분리에 대한 문제가 아니라 , 
도메인 적이지 않은 단순 유틸리티 성인 접근 때문에 질문을 하셨던 거군요!
이해 했습니다!

lxxjn0 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,52 @@
+        return Cars.from(nameCatalog);
+    }
+
+    private void racing(final List<Car> cars, Integer raceCount){
```
메서드는 행위를 하는 코드인데 이름이 명사가 사용되고 있어요. 행위를 드러낼 수 있는 이름으로 바꿔보면 좋을 것 같습니다.

youngsu5582 replied:
네이밍은 언제나 어려운거 같습니다,,,

요구사항 or 개발을 모르는 사람이 봐두 이해하기 좋은 이름으로 수정해보겠습니다!

lxxjn0 commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,36 @@
+        //Then
+        assertInstanceOf(Car.class, car);
+    }
+    @ParameterizedTest
+    @DisplayName("4 이상이면 위치를 한 칸 전진한다.")
+    @ValueSource(ints={4,5})
+    public void forwardPositionWhenNumberIsGreaterThan4(Integer value){
+        Car car = new Car(new Name("kim"));
+        Integer pos=car.getPosition();
+
+        car.race(value);
+
+        assertEquals(car.getPosition(),pos+1);
+    }
```
전진하는 케이스에 대한 테스트를 작성하였다면 전진하지 않는 케이스에 대한 테스트도 작성해보는 것이 어떨까요?

youngsu5582 replied:
앗차차 까먹었습니다... ( 핑계 아닙니다 😞 )
근데 해당 부분에 대해서 추가적으로 
항상 궁금증을 가졌던게 이렇게 숫자의 값에 따라 결과가 나뉘는 로직이 있으면
테스트 코드에 해당 숫자를 구체적으로 드러내는지 , 드러내지 않는지에 대해 궁금합니다.

```java
@Test
@DisplayName("파워가 일정 크기 이상이면 한 칸 전진한다.")
public void forward(){

}
@Test
@DisplayName("파워가 4 이상이면 한 칸 전진한다.")
public void forward(){   
}
```
제 개인적인 의견으로는
테스트는 보자마자 해당 기능에 대한 내용과 사용법을 명확하게 보여줘야 한다고 생각합니다

일정 크기 이상으로 명시를 하면 forward 함수를 클릭해서 한번더 봐야만 알수있습니다.
특히 , 값들을 상수에 분리하거나 보여주지 않으면 더 확인하기 힘들다고 생각합니다

그렇기에 , 해당 값이 바뀌면 DisplayName 을 변경하는 한이 있더라도 명시를 해주는게 좋다고 생각합니다.
제 의견이 맞는지나 , 스티치는 어떻게 생각하는지에 대해 피드백 주시면 감사하겠습니다! 🙏🙏

lxxjn0 replied:
저는 DisplayName에 이름을 드러내는 것이 맞다고 생각합니다. 어떠한 내용을 테스트하는지 확실히 드러날 수 있기 때문이죠!

추가로 아래 이미지를 참고하신다면 위 고민을 좀 더 간단하게 해결할 수도 있을 것 같아 공유드립니다.

<img width="764" alt="image" src="https://github.com/woowacourse/java-racingcar/assets/48052622/1678b5c6-5e51-4a65-aa4b-c944d4e2e4f9">

lxxjn0 commented on src/test/java/domain/NameTest.java:
```diff
@@ -0,0 +1,44 @@
+        //Then
+        assertInstanceOf(Name.class,name);
+    }
+    @ParameterizedTest
+    @DisplayName("공백 이름은 에러를 발생시킨다")
+    @ValueSource(strings = {""," "})
+    public void throwExceptionWhenValueIsBlank(String value){
+        assertThrows(IllegalArgumentException.class,()->{
+            new Name(value);
+        });
+    }
```
ParameterizedTest를 활용해서 테스트 로직을 구현하였네요 👍🏼

lxxjn0 reviewed (changes requested):
안녕하세요, 조이썬 :)
이번에 리뷰를 맡게 된 스티치라고 합니다.

자동차 게임에서 최대한 도메인을 분리하면서 고민을 한 노력이 보이는 코드였습니다 👍🏼 

코드의 전반적인 부분에 대해서 피드백을 드리고는 싶지만, 그러기 보다는 자동차 미션에서 숙지하고 넘어가야 할 중요한 부분에 대해서 좀 더 집중적으로 리뷰를 드리기로 하였습니다. 그래서 생각했던 것 보다는 피드백이 조금은 쉬워보일 수도 있습니다.

그래도 쉬운 것이 중요하지 않은 것은 아니기에, 잘 확인하고 고민해보고 피드백 반영 부탁드리겠습니다 :)

추가로 반복적으로 피드백이 필요한 부분에 대해서는 따로 코멘트를 더 남기지는 않았습니다. 제가 작성한 피드백은 전체 코드에 해당하는 피드백이라고 생각하고 코드 전체적으로 반영해주시면 좋을 것 같아요 😃

---

코드 전반적으로 코딩 컨벤션이 적용되지 않은 것으로 보여집니다. 구글 자바 코드 컨벤션을 한번 공부해보시고 코드 전체적으로 컨벤션 적용 부탁드리겠습니다.

입출력과 관련된 비즈니스 로직이 도메인 로직에 포함되어 있는 경우가 보입니다(toString 메서드). 조이썬이 생각하고 가능하다고 생각하는 범위 내에서 최대한 입출력 로직을 도메인 로직에서 분리해보면 좋을 것 같아요 :)

그리고 앞서 여쭤보신 질문에 대해서 간단하게 답변 드려보겠습니다.

### Controller, Service 기타 등등

이 부분에 대해서는 현재는 고려하지도, 고민하지도 않으셔도 될 것 같습니다. 지금처럼 그냥 도메인 객체와 출력 객체만 고려하시고 작성하셔도 괜찮을 것 같아요 :)

### 일급 객체 / 일급 컬랙션

코드 전반적으로 객체를 분리하려고 노력하신 모습이 보였던 것 같습니다. 말씀하신 '단순히 리스트를 감싸기 위한 일급 객체(컬랙션)이 필요한가?'란 질문에 대한 답을 하려면, 그 전에 왜 일급 객체와 일급 컬랙션이 필요한가에 대한 조이썬의 생각이 중요할 것 같습니다.

일급 객체로 나누기 위한 일급 객체/컬렉션은 큰 의미가 없다고 생각합니다. 일급 객체나 컬렉션이 아니어도 코드는 충분히 작성할 수 있고 돌아가게 할 수 있습니다. 코드를 위한 코드를 작성하는 것이 아닌 개발자 조이썬의 생각과 기준이 담긴 코드가 작성되면 더 좋을 것 같습니다. 현재 코드는 많은 고민을 했다는 것은 느껴지지만 그 명확한 기준이나 근거는 보이지 않는 것 같아요.

이번에 피드백을 반영하면서 일급 객체와 컬랙션을 왜 쓰는지에 대해 한번 공부해보시는 것을 추천드리겠습니다. 그런 이후 미션에서 조이썬이 작성한 코드에서는 어떤 도메인까지 일급 객체이고 컬렉션이면 좋을지 고민해보고 개선해보신 후 다시 이 질문에 대해 이야기해보면 좋을 것 같습니다.

### 단순 객체

조이썬이 예시로 든 코드는 입출력을 위한 매개체 객체라고 볼 수 있을 것 같아요. 이러한 객체를 잘 활용한다면 입출력 로직과 도메인 로직의 의존성을 끊어서 영향 범위를 제한 할 수 있습니다. 

그러나 코드에서 이 객체를 적절하게 사용하였나에 대해서는 그렇지 않았던 것 같습니다. 현재는 굳이 이런 객체를 쓸 필요는 없다고 생각합니다. 먼저 이런 매개체 역할을 하는 객체없이 개발을 해보시고 '도메인 로직과 입출력 로직의 의존성을 끊기가 너무 어려운데?'라는 생각이 드실 때 다시 고민해보시길 추천드리겠습니다 👍🏼

youngsu5582 commented:
스티치 답변이 너무 늦어서 죄송합니다..! 
너무 좋은 내용들 이기에
질문을 하나씩 보고 답변을 생각(음미...?) 했습니다

정성스런 리뷰와 질문들 너무 감사하고
앞으로의 과제들과 제가 코드를 짤때 
계속 유의해야할 기본기 적인 사항 많아서 너무 좋았습니다! ☺️☺️

내일 저녁까지 최대한 리팩토링 다 끝내고 다시 리뷰 요청 올리겠습니다!

youngsu5582 committed:
refactor : NamesCatalog -> Names 로 변경

- Catalog 역시 이름목록을 나타내기에 부적합 하다가 판단하여 Names로 변경
- 이에 따른 코드들 수정

youngsu5582 committed:
refactor : RaceResult 삭제 , Cars 에서 로직 담당

- RaceResult 를 오버 엔지니어링으로 판단하여 삭제
- Cars 에서 해당 도메인이 가진 기능 수행

- 이에 따른 테스트 코드 역시 반영 ( 삭제 / 코드 추가 )

youngsu5582 committed:
refactor : Names validate 부분 Name equals,hashCode 오버라이딩으로 해결

- Names 의 validate 부분 set 로직 간결화
- Name 의 equals,hashCode 재정의로 value 일치 여부로 재정의

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1491133683)

youngsu5582 committed:
test(Car) : 차가 전진하지 않는 경우에 대한 테스트 코드 추가

4 미만이면 위치를 전진하지 않는다 케이스 추가

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1491160619)

youngsu5582 committed:
refactor(Car) : Car 가 가지고 있는 FORWARD_MAX_NUMBER 상수 파일로 분리

- Car 의 FORWARD_MAX_NUMBER 제거
- CarConstant 에 MAXIMUM_POWER 추가

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1491136070)

youngsu5582 committed:
refactor : Domain 에 있는 입출력 담당 toString OutputView 로 분리

- RaceCarInfo , Car 부분의 toString 제거
- OutputView 에 입출력 로직 추가

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1491138317)

youngsu5582 committed:
refactor : OutputView 에 있는 하드코딩 상수로 분리

- " : " INFO_SEPARATOR 로 분리
- "-" POSITION_INDICATOR 로 분리

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1491148604)

youngsu5582 committed:
refactor : NumberGenerator 제거 -> GameController 에서 담당

- NumberGenerator 제거
- GameController createPower 에서 담당

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1491154329)

youngsu5582 committed:
refactor : GameController 내 메소드 명 변경

- register -> registerCars
- racing -> performRace
- race -> performEachRace
- finish -> announceWinners

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1491159314)

youngsu5582 committed:
refactor(all) : Intellij 자동 정렬 통한 코드 스타일 준수

youngsu5582 committed:
refactor : Car 에서 직접 위치 비교하게 구현

- Car 에서 isAtPosition 통해 비교

youngsu5582 committed:
refactor : Names 가 불변 List 반환하게 변경 , 일급 컬렉션 값 value 로 통일

youngsu5582 committed:
docs : 기능 요구사항 재수정

단순 코드 명세서가 아닌 , 기능 요구사항에 맞게 재수정

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1491110411)

youngsu5582 commented:
너무 늦은 시간에 코멘트 남긴거 같아서 죄송합니다 스티치...! 🍎🍎 😭

코드를 리팩토링 해가며 기본기만 부족함을 참 많이 느낀거 같습니다

특히
일급 객체나 일급 컬렉션의 차이점 이나 왜사용해야하는지
toString 을 어떨때 사용해야하는지
입출력 과 도메인 로직의 분리

등등 다양한 면에서 다시 한번 공부한 거 같아 참 좋았습니다

질문들에 대한 답변은 간단하게 conversation 으로 남겨놓았습니다
-> 해당 답변들에 대해서도 피드백 해주시면 완전 감사하겠습니다! ( 꼬리 질문도 완전 좋습니다!!! )

해당 코드를 다시 리팩토링 해가며 느낀 추가적인 점들에 대해서 조금만 더 남기겠습니다 ,,,

#### Record Class

<img width="792" alt="image" src="https://github.com/woowacourse/java-racingcar/assets/98307410/63789da7-0380-444c-876e-edb745cb7c18">

이렇게 도메인 들에 대한 클래스들을 작성하다 보면 레코드로 변환을 할 수가 있다고 나오는 경우가 종종 있습니다.

<img width="792" alt="image" src="https://github.com/woowacourse/java-racingcar/assets/98307410/03955451-f583-47a2-8c4e-b8045b7acde2">

레코드를 통해 작성하면
1. 변수들이 자동 final + getter
2. 코드 간결화
3. equals & hashCode 자동 재정의

이러한 장점들이 있다고 공부 했었습니다.

도메인을 설계하는 과정에서도 레코드를 사용해도 되나요??
도메인 설계 단에서 레코드를 사용하면 안되는 경우나 이유가 있는지 궁금합니다!

없다면 레코드를 사용하는게 보일러 플레이트 작성 비용이나 가독성이 더 좋을거 같아서 입니다!

#### 일급 컬렉션 테스트

말씀해주신 피드백 덕분에 , 
Car 들을 가진 일급 컬렉션인 Cars 에서 기능을 하는게 더 좋다는 것을 깨달았습니다.

```java
package domain.car;

import domain.name.Names;

import java.util.Collections;
import java.util.List;

public class Cars {
    private final List<Car> value;

    public Cars(List<Car> cars) {
        this.value = cars;
    }

    public static Cars from(Names names) {

        List<Car> cars = names.getNames()
                              .stream()
                              .map(Car::new)
                              .toList();

        return new Cars(cars);
    }

    public List<Car> getMaxPositionCars() {
        Integer maxPosition = getMaxPosition();
        return selectCarsWithPosition(maxPosition);
    }

    private Integer getMaxPosition() {
        return value.
                stream()
                .mapToInt(Car::getPosition)
                .max()
                .orElse(0);
    }

    private List<Car> selectCarsWithPosition(Integer position) {
        return
                value.stream()
                     .filter(car -> car.isAtPosition(position))
                     .toList();
    }

    public List<Car> getValue() {
        return Collections.unmodifiableList(value);
    }
}
```

이렇게 Cars 내부에서 기능들을 구현하며 드는 생각은
getValue 가 필요없다 였습니다.

간단한 로직에선 값을 반환할 필요 없이
로직을 수행한 값만 반환하면 되기 때문입니다.

하지만 , 테스트 를 할때
```java
    @Test
    @DisplayName("자동차들중 가장 많이 간 우승 자동차들을 모은 경기결과를 만든다.")
    public void createRaceResult() {
        Name pond = new Name("pond");
        Name poby = new Name("poby");
        Names names = new Names(List.of(pond, poby));
        Cars cars = Cars.from(names);

        cars.getValue()
            .get(0)
            .race(5);
        List<Car> winnerCars = cars.getMaxPositionCars();

        assertEquals(winnerCars.size(), 1);
        assertEquals(winnerCars.get(0)
                               .getName(), pond);
    }
```
getValue 를 통해 값을 조작한 후에
검증을 해야만 했습니다.

```java
    Car(Name name,int position){}
```
이렇게 생성자를 추가적으로 만들어 테스트가 가능하나
이 역시도 getValue 와 같이 테스트를 위한 코드가 만들어 지는거 같았습니다.

이런 경우 어떻게 해야할까요..?
( 추가로 , 이런 일급 컬렉션일때 변수명은 어떻게 하는지 궁금합니다..!
  저는 변수가 하나밖에 없고 , 클래스 명이 Cars 이렇게 Car 를 가짐을 명시해주므로 value 라고 사용해도 괜찮을 거 같다고 생각합니다.! )

다시 한번 좋은 피드백 너무 감사합니다!!
덕분에 , 레벨 1 리팩토링을 정말 알차고 효율있게 진행한거 같아요!! 🫡🫡

lxxjn0 commented on src/main/java/domain/car/Car.java:
```diff
@@ -2,14 +2,14 @@
     private final Name name;
-    private Integer position;
-    private static final Integer FORWARD_THRESHOLD=4;
+    private int position;
+    private static final Integer FORWARD_POWER_THRESHOLD = 4;
```
자바 코딩 컨벤션을 지켜주세요.

lxxjn0 commented on src/main/java/controller/GameController.java:
```diff
@@ -1,52 +1,64 @@
+    private void performEachRace(final List<Car> cars) {
         cars.forEach(car -> {
-            car.race(NumberGenerator.generate(Car.FORWARD_MAX_NUMBER));
+            Integer power = createPower();
```
[질문] 해당 라인에서 자료형을 참조형으로 사용하신 이유가 따로 있을까요? 혹시 다른 이유가 있는 것이 아니라면 Java의 primitive type과 wrapper class에 대한 키워드로 한번 학습해보시면 좋을 것 같습니다.

youngsu5582 replied:
없습니다,,
습관성 Integer 사용 이였던 거 같습니다 😥😥

lxxjn0 commented on src/test/java/domain/CarTest.java:
```diff
@@ -22,15 +22,29 @@ void createCar(){
+    @DisplayName("4 미만이면 위치를 전진하지 않는다")
+    @ValueSource(ints = {0, 3})
+    public void notForwardPosition(int value) {
+        Car car = new Car(new Name("스티치"));
```
ㅋㅋㅋㅋ 저도 코드 한켠에 출연하였네요 😄

lxxjn0 commented on src/main/java/view/OutputView.java:
```diff
@@ -1,28 +1,58 @@
     }
-    public void printRaceProgresses(List<RaceProgress> raceProgresses){
+
+    public void printAllRaceProgress(List<RaceProgress> raceProgresses) {
```
입출력 로직에 출력과 관련된 비즈니스를 모아둠으로써 도메인 로직과 입출력 로직의 의존성을 끊을 수 있었네요 👍🏼

lxxjn0 reviewed (approved):
조이썬, 피드백 반영하시느라 수고 많으셨습니다 :)

앞선 리뷰에 대한 답변을 추가로 남겨두었으니 참고하셔서 2단계 진행해주시면 될 것 같아요.
이어서 추가로 질문주신 내용에 대한 답변입니다.

---

### Record class

제가 주로 Java 8만 사용하다보니 Record 클래스에 대한 이해는 부족할 수 있으나 Kotlin의 Data 클래스와 같다고 알고 있어서 해당 내용을 기반으로 말씀드리겠습니다.

제가 알기로 Record 클래스는 데이터 클래스로써 활용하기 위해 나온 것으로 알고 있습니다. 객체의 프로퍼티는 변경이 불가능한 불변으로만 생성이 가능하고, 이는 흔히 DTO의 용도를 생각해주시면 될 것 같아요. 물론 불변 객체로의 사용이 가장 보편적인 영역이 DTO라서 예시로 든 것이지 도메인 클래스에서 Record를 사용하면 안된다는 것은 아닙니다. 조이썬이 생각하기에 도메인 클래스이면서 불변으로 관리해야 하는 부분이라면 당연히 사용하셔도 됩니다.

추가로 도메인 객체로 관리하는 대상은 보통 속성(필드)을 가지며, 이 속성을 관리하고 변화시키면서 도메인 로직을 전개해나갑니다. 그렇기에 단순히 불변으로 관리되는 객체라면 도메인 객체로의 사용 용도가 적절한가?를 먼저 고민해볼 수 있을 것 같아요. 물론 불변이라고 도메인 객체가 아니란 이야기는 아니니깐 이 부분을 고려 포인트로 가져가면 좋을 것 같다는 이야기를 드리고 싶어서 남긴 이야기니 참고만 해주세요.

### 일급 컬렉션

해당 부분은 제가 이전 리뷰에 남긴 추가 답변이 대답이 될 수 있어서 따로 작성하진 않도록 하겠습니다. 그러나 혹시 이해가 가지 않거나 더 궁금한 부분이 있다면 언제든 질문주세요 :)
